internal/datastore/mysql/datastore.go

```diff
@@ -0,0 +1,648 @@
+func (mds *Datastore) collectGarbageForTransaction(ctx context.Context, highest uint64) (int64, int64, error) {
+	// Delete any relationship rows with deleted_transaction <= the transaction ID.
+	relCount, err := mds.batchDelete(ctx, mds.driver.RelationTuple(), sq.LtOrEq{colDeletedTxn: highest})
+	if err != nil {
+		return 0, 0, err
+	}
+
+	log.Trace().Uint64("highestTransactionId", highest).Int64("relationshipsDeleted", relCount).Msg("deleted stale relationships")
+	// Delete all transaction rows with ID < the transaction ID. We don't delete the transaction
+	// itself to ensure there is always at least one transaction present.
+	transactionCount, err := mds.batchDelete(ctx, mds.driver.RelationTupleTransaction(), sq.Lt{colID: highest})
+	if err != nil {
+		return relCount, 0, err
+	}
+
+	log.Trace().Uint64("highestTransactionId", highest).Int64("transactionsDeleted", transactionCount).Msg("deleted stale transactions")
+	return relCount, transactionCount, nil
+}
+
+// FIXME dupe from postgres datastore - need to refactor
+// - query was reworked to make it compatible with Vitess
+// - API differences with PSQL driver
+func (mds *Datastore) batchDelete(ctx context.Context, tableName string, filter sqlFilter) (int64, error) {
+	query, args, err := sb.Delete(tableName).Where(filter).Limit(batchDeleteSize).ToSql()
+	if err != nil {
+		return -1, err
+	}
+
+	var deletedCount int64
+	for {
+		cr, err := mds.db.ExecContext(ctx, query, args...)
+		if err != nil {
+			return deletedCount, err
+		}
+
+		rowsDeleted, err := cr.RowsAffected()
+		if err != nil {
+			return deletedCount, err
+		}
+		deletedCount += rowsDeleted
+		if rowsDeleted < batchDeleteSize {
+			break
+		}
+	}
+
+	return deletedCount, nil
+}
+
+// IsReady returns whether the datastore is ready to accept data. Datastores that require
+// database schema creation will return false until the migrations have been run to create
+// the necessary tables.
+//
+// fundamentally different from PSQL implementation:
+// - checking if the current migration version is compatible is implemented with IsHeadCompatible
+// - Database seeding is handled here, so that we can decouple schema migration from data migration
+//   and support skeema-based migrations.
+func (mds *Datastore) IsReady(ctx context.Context) (bool, error) {
+	if err := mds.db.PingContext(ctx); err != nil {
+		return false, err
+	}
+
+	currentMigrationRevision, err := mds.driver.Version()
+	if err != nil {
+		return false, err
+	}
+
+	compatible, err := migrations.Manager.IsHeadCompatible(currentMigrationRevision)
+	if err != nil {
+		return false, err
+	}
+	if !compatible {
+		return false, nil
+	}
+
+	// seed base transaction if not present
+	txRevision, err := mds.HeadRevision(ctx)
+	if err != nil {
+		return false, err
+	}
+	if txRevision == datastore.NoRevision {
+		baseRevision, err := mds.seedBaseTransaction(ctx)
+		if err != nil {
+			return false, err
+		}
+		if baseRevision != datastore.NoRevision {
```

Comment:
remove and place log into seedBaseTransaction

Reply:
addressed in https://github.com/authzed/spicedb/pull/525/commits/d4f34e67559e70d540a1470fa8ce39869cfdf4de
I also realized the logic was scattered and that there was no need for that, so I refactored the method